lib/preload.js

```diff
@@ -361,32 +360,67 @@ class Preloader extends EventEmitter {
 
 	_prepareErrorHandler() {
 		// Emit an error event if the python script exits with an error
-		this.container.wait().then((data) => {
-			if (data.StatusCode !== 0 && data.StatusCode !== EXIT_CAUSED_BY_SIGINT) {
-				const output = this.bufferedStderr.getData().toString('utf8').trim();
-				let error;
-				if (
-					output.indexOf(GRAPHDRIVER_ERROR) !== -1 &&
-					output.indexOf(DOCKERD_USES_OVERLAY) !== -1
-				) {
-					error = new this.balena.errors.BalenaError(OVERLAY_MODULE_MESSAGE);
-				} else {
-					error = new Error(output);
-					// @ts-ignore
-					error.code = data.StatusCode;
+		this.container
+			.wait()
+			.then((data) => {
+				if (data.StatusCode !== 0) {
+					const output = this.bufferedStderr.getData().toString('utf8').trim();
+					let error;
+					if (
+						output.indexOf(GRAPHDRIVER_ERROR) !== -1 &&
+						output.indexOf(DOCKERD_USES_OVERLAY) !== -1
+					) {
+						error = new this.balena.errors.BalenaError(OVERLAY_MODULE_MESSAGE);
+					} else {
+						error = new Error(output);
+						// @ts-ignore
+						error.code = data.StatusCode;
+					}
+					this.emit('error', error);
 				}
-				this.emit('error', error);
-			}
-		});
+			})
+			.catch((error) => this.emit('error', error));
 	}
 
+	/**
+	 * Send a command to `preload.py` (running in a Docker container) through
+	 * its stdin, and read a response from its stdout. The stdout response
+	 * must be a JSON object containing a `StatusCode` field (number / integer)
+	 * and an `Error` field (string). If `StatusCode` is the number zero, the
+	 * command execution is considered successful, otherwise it is assumed to
+	 * have failed and the returned promise is rejected with an error message
+	 * including the message provided in the `Error` field (if any).
+	 *
+	 * @param {string} command
+	 * @param {object} parameters
+	 */
 	_runCommand(command, parameters) {
 		return new Bluebird((resolve, reject) => {
+			const cmd = JSON.stringify({ command, parameters }) + '\n';
 			this.stdout.once('error', reject);
```

Comment:
A normal dockerode failure will fall into here with `StatusCode` and `Error` having some values. Those values will never be present in the `data` callback.

Reply:
My main concern in ensure that the CLI process terminates (no infinite spinning wheels), preferably with a helpful error message. If `reject` is called here, I think the rejection bubbles up to the CLI and the process terminates gracefully.

Are you suspecting that the error message would not be helpful?